Fix #295: Catch libpng warnings

cmb69 opened the pull request:
Uncaught libpng warnings will either be printed to stderr or suppressed,
neither of which is desirable. Therefore we add a libpng warning handler,
similar to the already existing error handler, but the warning handler must
not `longjmp`.

cmb69 commented:
Note that I consider this issue to be more a bug than an enhancement, so IMHO the patch should applied to GD-2.2 as well.

vapier commented on tests/png/bug00295.c:
```diff
@@ -0,0 +1,45 @@
+#include <string.h>
+
+
+#define EXPECTED_MSG "gd-png: libpng warning: iCCP: known incorrect sRGB profile"
```
the exact message depends upon the exact string libpng emits, and the exact code paths that libpng goes through.  do we really want to tie ourselves to that ?

cmb69 replied:
Thanks, good catch. Thinking about it, *this* test shouldn't be concerned about the message at all, and not even about GD_WARNING, but rather that there's no GD_ERROR.

vapier commented on tests/png/bug00295.c:
```diff
@@ -0,0 +1,45 @@
+#define EXPECTED_MSG "gd-png: libpng warning: iCCP: known incorrect sRGB profile"
+
+
+static void error_handler(int priority, const char *format, va_list args);
```
just move the func up here and you don't need the forward decl

cmb69 replied:
I'd welcome a general style guide for libgd! :-)

cmb69 commented:
I've made the requested improvements. 

I'd prefer to squash the commits immediately before merging the PR (if at all).

vapier commented on tests/png/bug00295.c:
```diff
@@ -0,0 +1,39 @@
+
+static void error_handler(int priority, const char *format, va_list args)
+{
+    char msg[256];
```
do we have to worry about 256 being too low ?

cmb69 replied:
AFAIK libpng makes no guarantee regarding the max length of the error messages, so, yes, we have to worry. Is `vsnprintf` portable enough for our purposes?

vapier replied:
imo, if it's in POSIX, it's fine.  if we find a system that lacks the function, we can figure out how to make the function available to it as a fallback.  but better to just wait for someone to complain than always try to make things work for a hypothetical setup that never materializes.

vsnprintf is in POSIX: http://pubs.opengroup.org/onlinepubs/9699919799/functions/vsnprintf.html

it shouldn't be hard to find replacements out there we can drop in (like [gnulib](https://www.gnu.org/software/gnulib/)).

cmb69 replied:
> imo, if it's in POSIX, it's fine. […]

Sounds good to me, thanks. I'll switch to `vsnprintf` ASAP.

vapier commented on tests/png/bug00295.c:
```diff
@@ -0,0 +1,39 @@
+    fclose(fp);
+    gdImageDestroy(im);
+
+    return gdNumFailures();
```
what guarantees that `error_handler` was called ?  should we have a global boolean like `called` that starts off `false` and `error_handler` sets to `true` ?

cmb69 replied:
There is no guarantee that `error_handler` will be called, and actually libpng < 1.6 wouldn't throw an error (and the behavior might change in a future version again; might be configurable). Therefore I think the test is okay: we're just checking that no `GD_ERROR` will ever be raised.

vapier replied:
can we do a version check ?  if it's less than 1.6, don't require it to be set, but if it's >=1.6, require it ?

can we feed it a more damaged image that'll more strongly guarantee libpng calls us ?

vapier replied:
to be clear, i'm not saying you must do this ... it'd just be a nice to have since the current test doesn't actually hard verify that we set the error handler.  if we broke things in the future, this test case wouldn't help catch regressions.  i.e. i'm pretty sure it'll "pass" if you build it against a version of gd that doesn't have your fixes here.

cmb69 commented:
@vapier I've finally managed to rewrite the regression test so that it actually tests that the error handler is called exactly once from `gdPngWarningHandler()` (ignoring additional calls from `gdPngErrorHandler()` which are follow up errors).

I've rebased onto master and squashed the commits. Okay to merge?

woho commented:
Could it be that the merge of this fix was forgotten? I cannot find the commit on the current master branch.
